serve/mlc_serve/model/paged_cache_manager.py

```diff
@@ -381,26 +369,19 @@ def get_max_new_tokens(self) -> int:
         It should return the result of `get_kv_cache_size` if there is
         no requests in the cache.
         """
-        if not self.allocated_decode_tokens:
+        if not self.token_counts:
             return len(self.free_blocks) * self.block_size
 
-        free_blocks_per_sequence = len(self.free_blocks) // len(
-            self.allocated_decode_tokens
-        )
+        free_blocks_per_sequence = len(self.free_blocks) // len(self.token_counts)
         remaining_blocks = len(self.free_blocks) - free_blocks_per_sequence * len(
-            self.allocated_decode_tokens
+            self.token_counts
         )
-
-        total_tokens = []
-
-        for seq_id, tokens in self.allocated_decode_tokens.items():
-            prompt_seq_id = get_prompt_sequence_id(seq_id.request_id)
-            prompt_tokens = self.allocated_prompt_tokens[prompt_seq_id]
-            total_tokens.append(prompt_tokens + tokens)
-
+        # For parallel sampling, the number of shared prompt tokens is divisible
+        # by self.block_size (since the remainers are copied to each sequence).
+        # So the following calculation does not overcount shared prompt tokens.
         remaining_tokens_in_last_block = [
             self.block_size - (tokens - 1) % self.block_size - 1
-            for tokens in total_tokens
+            for tokens in self.token_counts.values()
```

Comment:
@sunggg Please check that this change and the comments above make sense. This was where we had a bug today.

Reply:
> (since the remainers are copied to each sequence)
Can this be expensive? I think it is okay for now but want to know future action item.

Reply:
It should be very cheap, it only happens for parallel sampling requests after prefill https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/model/paged_cache_model.py#L446-L460

I think it's cheaper than filling of KV tensors into paged cache that happens in every prefill.